src/neuralnet/accrual/snapshot.h

```diff
@@ -0,0 +1,1202 @@
+                            "  Baseline: %" PRIu64, entry.m_height);
+
+                        m_heights.clear();
+                        break;
+                    case Action::REGISTER:
+                        LogPrint(LogFlags::ACCRUAL,
+                            "  Added: %" PRIu64, entry.m_height);
+
+                        m_heights.emplace_back(entry.m_height);
+                        break;
+                    case Action::DEREGISTER:
+                        LogPrint(LogFlags::ACCRUAL,
+                            "  Removed: %" PRIu64, entry.m_height);
+
+                        m_heights.pop_back();
+                }
+            } catch (const std::ios_base::failure& e) {
+                if (feof(file.Get())) {
+                    break;
+                }
+
+                throw;
+            }
+        }
+    }
+}; // AccrualSnapshotRegistry
+
+//!
+//! \brief Manages storage of accrual snapshots.
+//!
+//! TODO: Add snapshot pruning to reclaim disk space for very old snapshots.
+//! TODO: Add a way to rebuild the snapshots in case of file corruption, etc.
+//!
+class AccrualSnapshotRepository
+{
+public:
+    //!
+    //! \brief Initialize the accrual snapshot system.
+    //!
+    //! \return \c false if the snapshot system failed to initialize because of
+    //! an error.
+    //!
+    bool Initialize()
+    {
+        try {
+            fs::create_directory(SnapshotDirectory());
+        } catch (const std::exception& e) {
+            return error(
+                "%s: failed to create the accrual snapshot directory %s: %s",
+                __func__,
+                SnapshotDirectory().string(),
+                e.what());
+        }
+
+        return m_registry.Initialize();
+    }
+
+    //!
+    //! \brief Determine whether the node already stored a baseline accrual
+    //! snapshot.
+    //!
+    //! \return \c true if the node previously activated the accrual snapshot
+    //! system.
+    //!
+    bool HasBaseline() const
+    {
+        return m_registry.BaselineHeight() > 0;
+    }
+
+    //!
+    //! \brief Store a snapshot of accrual for each account as the baseline.
+    //!
+    //! \param height   Height of the block to associate with the snapshot.
+    //! \param accounts Research accounts to record accrual from.
+    //!
+    //! \return \c false when an error occurs while creating a snapshot.
+    //!
+    bool StoreBaseline(const uint64_t height, const ResearchAccountMap& accounts)
+    {
+        return m_registry.ResetBaseline(height) && Store(height, accounts);
+    }
+
+    //!
+    //! \brief Store a snapshot of accrual for each account to disk.
+    //!
+    //! \param height   Height of the block to associate with the snapshot.
+    //! \param accounts Research accounts to record accrual from.
+    //!
+    //! \return \c false when an error occurs while creating a snapshot.
+    //!
+    bool Store(const uint64_t height, const ResearchAccountMap& accounts)
```

Comment:
Note that in the scraper I hash the file on save to and store the hashes in the ScraperFileManifestMap and Manifest.csv.gz. Should you do this too? What happens if the file is tampered with or is otherwise corrupted? I also don't see any true-up between the registry and the actual files on disk. I haven't thought through in detail with this code what happens if a file contained in the registry gets corrupted or deleted.

Reply:
Yes, good eye. I'm working on this now 🙂 

For the draft, I left the registry "dumb" while I work on getting those internals figured out. The registry is supposed to provide an additional layer of integrity for the snapshot files, resolve discrepancies on disk, and allow for cleaning up old snapshots. Currently, it doesn't do much besides identifying the files on disk.

Reply:
What about storing this in leveldb rather than files?
The reason I used the files for the scraper is that they are essentially one-to-one with the files at the source (project). But do you think leveldb is more robust?

Reply:
By the way. I am perfectly fine with using individual files for this, as long as the ability of the wallet to detect a missing/corrupted file and self heal is there.

Sometimes people implement database storage when in fact file storage can be extremely reliable if implemented in a reasonably robust way. The scrapers for example run for MONTHS without any trouble...

Reply:
We could use LevelDB, and we probably will when we get to the point where we need to drop some of the research reward accounting out of memory to scale. For the snapshots, we're not storing many records, there are no complex access/modification patterns, and the data is trivially structured--as you said, a database doesn't provide much of an advantage over the flat files and adds a bit of overhead for this use case. I'm also not especially familiar with LevelDB's API, limitations, and best practices, so I chose file-backed storage for simplicity in the meantime. We can upgrade the underlying storage mechanism without a mandatory change to the protocol later if needed.

Reply:
I think I am good with the flat files. As long as the wallet doesn't malfunction due to a corrupted or missing file, I think it is ok.

Reply:
@cyrossignol what happens right now if a snapshot file is ripped out after the SnapshotBaselineBuilder has been run?

Reply:
It is very clear this system will work, and will generate deterministic accruals that will be consistent across all nodes, as long as it is resilient to the destruction or corruption of a snapshot file.

Reply:
The node only needs accrual snapshots on start-up and when reorganizing across a superblock. If a reorg never happens for the height of a deleted/corrupted snapshot file, the node will continue to run normally until the next superblock when it creates a new snapshot. If the node restarts before then, it cannot apply the latest snapshot and initialization fails. 

In the current draft, this is fatal. A node will need to resync the chain just like it would if something corrupts any of the other blockchain files. Since snapshot data is just rolled up from data in the chain, we can regenerate it, and I'm planning to add this. It's more for convenience. 

More importantly, I need to finish adding integrity checks for loading snapshots so that a node avoids using invalid data.

Reply:
Yes. Putting hash validation in the registry is the right answer. Can you just reinitialize rather than requiring a full resync?

Reply:
I will submit changes for snapshot file integrity in a separate PR. We decided to review as-is to expedite testing.

Reply:
Agreed.